### PR TITLE
[Codegen] Add pattern for hoisting scf.forall from scf.for

### PR DESCRIPTION
**Pull request by qedawkins**
This pattern does what is effectively subset hoisting on an scf.forall perfectly nested inside of an scf.for op, updating the iter arg of the for loop to use the slice read by and written by the parallel loop.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 1]
```diff
@@ -174,6 +176,225 @@ void transform_dialect::ApplyFoldFillIntoPadPatternsOp::populatePatterns(
...
+struct HoistForallFromFor : public OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(scf::ForOp loop,
+                                PatternRewriter &rewriter) const final {
+    if (loop.getBody()->getOperations().size() != 2) {
+      return rewriter.notifyMatchFailure(
+          loop, "Body of the loop contains more than one op");
+    }
+
+    auto forallOp =
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:196]: To restrictive? Maybe just look for a single forall op in the body of the outer for?

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 2]
```diff
@@ -174,6 +176,225 @@ void transform_dialect::ApplyFoldFillIntoPadPatternsOp::populatePatterns(
...
+      return rewriter.notifyMatchFailure(
+          loop, "Slice operand producers not safe to hoist out of loop");
+    }
+
+    // Sort the backwards slice of the producers for the insertion/extraction
+    // indices by the block order in the scf.forall body. This ensures that we
+    // hoist operations in the same order they started. Any topological ordering
+    // would work too because the operations are speculatable.
+    SmallVector<Operation *> sliceVec(slice.getArrayRef());
+    std::sort(sliceVec.begin(), sliceVec.end(),
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp]: Just use the`mlir::topologicalSort` routine?

**Review by MaheshRavishankar — APPROVED** [review]
That isnt too bad actually... Looks good to me.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp (hunk 2 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp]: oh cool, I didn't know this existed

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp (hunk 1 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:196]: I wanted to slow walk it a bit. Allowing other operations in the loop means we will be moving "block level code" into a thread level context, which I'd prefer to be prescriptive about.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:196]: I think there is no difference. Outside of the scf.forall its all thread level code (all threads are doing that redundantly). But I see your point.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:196]: You're probably right. I'm going to land this as is and address this in a follow up if it becomes an issue then.

**Pull request merged**